File: lib-utils/src/main/java/com/cqteam/lib/utils/SP.java
package com.cqteam.lib.utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.ArrayMap;

import com.cqteam.lib.UtilsManager;

/**
 * Created by koloces on 2020/5/8
 */
public class SP {
    private static SharedPreferences spf;
    private static ArrayMap<String, SharedPreferences> spfs;
    private SP() {
    }
    public static void init() {
        spf = UtilsManager.INSTANCE.getContext().getSharedPreferences("publicSP", Context.MODE_PRIVATE);
    }

    public static SharedPreferences getTagSpf(String tag) {
        if (StringUtils.INSTANCE.isEmpty(tag)) {
            return spf;
        }
        if (spfs == null) {
            spfs = new ArrayMap<>();
        }
        SharedPreferences sharedPreferences = spfs.get(tag);
        if (sharedPreferences == null) {
            sharedPreferences = UtilsManager.INSTANCE.getContext().getSharedPreferences(tag, Context.MODE_PRIVATE);
            spfs.put(tag, sharedPreferences);
        }
        return sharedPreferences;
    }

    public static void put(String key, Object obj) {
        put(null, key, obj);
    }

    /**
     * 通过 instanceof 来判断数据类型再强转成对应的数据类型进行存储
     *
     * @param key
     * @param obj
     */
    public static void put(String tag, String key, Object obj) {
        if (obj == null) {
            return;
        }
        SharedPreferences.Editor editor = getTagSpf(tag).edit();
        String simpleName = obj.getClass().getSimpleName();

        if (obj instanceof String || "String".equals(simpleName)) {
            editor.putString(key, (String) obj);
        } else if (obj instanceof Long || "Long".equals(simpleName)) {
            editor.putLong(key, (Long) obj);
        } else if (obj instanceof android.R.integer || "Integer".equals(simpleName)) {
            editor.putInt(key, (int) obj);
        } else if (obj instanceof Boolean || "Boolean".equals(simpleName)) {
            editor.putBoolean(key, (Boolean) obj);
        } else if (obj instanceof Float || "Float".equals(simpleName)) {
            editor.putFloat(key, (Float) obj);
        }
        editor.apply();
    }

    /**
     * 通过instanceof 来判断数据类型再进行取值然后再进行类型转换，这种方法比较麻烦取出值之后在应用的地方还要进行类型转换
     *
     * @param key
     * @param obj
     * @return
     */
    public static Object get(String tag, String key, Object obj) {
        if (obj instanceof String) {
            return getTagSpf(tag).getString(key, (String) obj);
        } else if (obj instanceof String) {
            return getTagSpf(tag).getLong(key, (Long) obj);
        } else if (obj instanceof android.R.integer) {
            return getTagSpf(tag).getInt(key, (int) obj);
        } else if (obj instanceof Boolean) {
            return getTagSpf(tag).getBoolean(key, (Boolean) obj);
        } else if (obj instanceof Float) {
            return getTagSpf(tag).getFloat(key, (Float) obj);
        }

        return null;
    }

    public static Object get(String key, Object obj) {
        return get(null, key, obj);
    }

    /**
     * 全部清除文件中的内容
     */
    public static void clearSpf(String tag) {
        SharedPreferences.Editor editor = getTagSpf(tag).edit();
        editor.clear().commit();
    }

    public static void clearSpf() {
        clearSpf(null);
    }

    /**
     * 清除指定key 的内容
     *
     * @param key
     */
    public static void remove(String tag, String key) {
        SharedPreferences.Editor editor = getTagSpf(tag).edit();
        editor.remove(key).commit();
    }

    public static void remove(String key) {
        remove(null, key);
    }


    /**
     * 获取String值
     *
     * @param key
     * @param defValue
     * @return
     */
    public static String getString(String tag, String key, String defValue) {
        return getTagSpf(tag).getString(key, defValue);
    }

    public static String getString(String key, String defValue) {
        return getString(null, key, defValue);
    }

    public static String getString(String key) {
        return getString(null, key, "");
    }

    /**
     * 获取int值
     *
     * @param key
     * @param defValue
     * @return
     */
    public static int getInt(String tag, String key, int defValue) {
        return getTagSpf(tag).getInt(key, defValue);
    }

    public static int getInt(String key, int defValue) {
        return getInt(null, key, defValue);
    }

    /**
     * 获取long值
     *
     * @param key
     * @param defValue
     * @return
     */
    public static long getLong(String tag, String key, long defValue) {
        return getTagSpf(tag).getLong(key, defValue);
    }

    public static long getLong(String key, long defValue) {
        return getLong(null, key, defValue);
    }

    /**
     * 获取float值
     *
     * @param key
     * @param defValue
     * @return
     */
    public static float getFloat(String tag, String key, float defValue) {
        return getTagSpf(tag).getFloat(key, defValue);
    }

    public static float getFloat(String key, float defValue) {
        return getFloat(null, key, defValue);
    }

    /**
     * 获取bo0lean值
     *
     * @param key
     * @param defValue
     * @return
     */
    public static boolean getBoolean(String tag, String key, boolean defValue) {
        return getTagSpf(tag).getBoolean(key, defValue);
    }

    public static boolean getBoolean(String key, boolean defValue) {
        return getBoolean(null, key, defValue);
    }
}

File: lib-utils/src/main/java/com/cqteam/lib/utils/sp/SP.kt
package com.cqteam.lib.utils.sp

import android.content.Context

/**
 *
 * @Description:    SharedPreferences的管理类，get,put为默认的，
 * 如果有定制的，比如用户信息等等，调用[getSpHelper],用户清楚时可以直接清楚
 * @Author:         koloces
 * @CreateDate:     2020/10/22 10:55
 */
object SP {
    val sps: MutableMap<String, SPHelper> by lazy {
        mutableMapOf<String, SPHelper>()
    }
    private var defaultSp: SPHelper? = null
    private val DEFAULT = "default"
    private var mContext: Context? = null

    fun init(context: Context) {
        mContext = context
        defaultSp = SPHelper(context, DEFAULT)
    }

    fun put(key: String?, obj: Any?) {
        defaultSp?.put(key, obj)
    }

    fun putStringList(key: String, list: List<String>) {
        defaultSp?.putStringList(key, list)
    }

    fun clean() {
        defaultSp?.clean()
    }

    fun remove(key: String) {
        defaultSp?.remove(key)
    }

    fun getString(key: String, default: String = ""): String {
        return defaultSp?.getString(key, default) ?: ""
    }

    fun getInt(key: String, default: Int): Int {
        return defaultSp?.getInt(key, default) ?: default
    }

    fun getBoolean(key: String, default: Boolean): Boolean {
        return defaultSp?.getBoolean(key, default) ?: default
    }

    fun getFloat(key: String, default: Float): Float {
        return defaultSp?.getFloat(key, default) ?: default
    }

    fun getLong(key: String, default: Long): Long {
        return defaultSp?.getLong(key, default) ?: default
    }

    fun getStringList(key: String): List<String> {
        return defaultSp?.getStringList(key) ?: mutableListOf()
    }

    /**
     * 获取特定的SPHelper
     */
    fun getSpHelper(tag: String): SPHelper {
        if (mContext == null) {
            throw SPNotInitException()
        } else {
            var spHelper = sps[tag]
            if (spHelper == null) {
                spHelper = SPHelper(mContext!!, tag)
            }
            sps[tag] = spHelper
            return spHelper
        }
    }

    fun getDefaultSPHelper(): SPHelper {
        if (mContext == null) {
            throw SPNotInitException()
        }
        return defaultSp!!
    }

    private class SPNotInitException : Exception("UtilsManager Not init")
}